db/view/view_update_generator.cc

```diff
@@ -40,14 +40,21 @@ future<> view_update_generator::start() {
                 _pending_sstables.wait().get();
             }
 
-            // To ensure we don't race with updates, move the entire content
-            // into a local variable.
-            auto sstables_with_tables = std::exchange(_sstables_with_tables, {});
-
-            // If we got here, we will process all tables we know about so far eventually so there
-            // is no starvation
-            for (auto table_it = sstables_with_tables.begin(); table_it != sstables_with_tables.end(); table_it = sstables_with_tables.erase(table_it)) {
-                auto& [t, sstables] = *table_it;
+            // We process a small batch from each table in turn to ensure fair incremental progress.
+            for (auto table_it = _sstables_with_tables.begin(); table_it != _sstables_with_tables.end();) {
+                auto& [t, all_sstables] = *table_it;
+                // Normally, we expect sstables to be added in order, so this
+                // should be a noop. It can help however when scylla is restarted
+                // with staging sstables, as they are not loaded from disk in order.
+                std::sort(all_sstables.begin(), all_sstables.end(), [] (const sstables::shared_sstable& a, const sstables::shared_sstable& b) {
+                    return a->get_first_decorated_key().token() < b->get_first_decorated_key().token();
+                });
+                std::vector<sstables::shared_sstable> excluded_sstables(all_sstables.begin(), all_sstables.end());
+                std::vector<sstables::shared_sstable> sstables;
+                while (!all_sstables.empty() && sstables.size() < sstable_batch_size) {
+                    sstables.push_back(std::move(all_sstables.front()));
```

Comment:
if push_back() throws, there's no effect on all_sstables, right? but we can't do rewind either, so probably good to reserve sstable_batch_size in sstables upfront.

Reply:
Good point, I'll use reserve to make sure that doesn't happen.

Reply:
I made the change locally, just forgot to push the branch.